java/fury-core/src/main/java/org/apache/fury/resolver/ClassResolver.java

```diff
@@ -391,12 +391,7 @@ private void registerDefaultClasses() {
 
   /** register class. */
   public void register(Class<?> cls) {
-    if (!extRegistry.registeredClassIdMap.containsKey(cls)) {
-      while (extRegistry.registeredId2Classes.containsKey(extRegistry.registeredClassIdCounter)) {
-        extRegistry.registeredClassIdCounter++;
-      }
-      register(cls, extRegistry.registeredClassIdCounter);
-    }
+    register(cls, extRegistry.registeredClassIdCounter);
```

Comment:
What will happen if there are some hold in `registeredId2ClassInfo`? 
For example:
```java
// registeredClassIdCounter = 99
fury.register(class1); // registeredClassIdCounter = 100 now
fury.register(class2, 101); // registeredClassIdCounter = 101 now
fury.register(class3); // throw exception since 101 has been used
```

Reply:
The situation you mentioned above cannot be covered because the user can specify the classId.

Duplicate registration of ClassID will be detected.

Reply:
We may be able to provide an interface to users in Fury, for example named `allocateClassId`, so that users can easily use ClassID to prevent the provided ClassID from being registered repeatedly (because the user does not know the allocation of ClassId within ClassResolver, so the ClassID provided by the user There is also the risk of being registered. ClassID is only maintained by the system, and obtaining ClassID is transparent to users).

What about this idea ?

Reply:
Yes, registered class id should be exposed to users, so they can know which ids they could use to avoid the conflict

Reply:
> What will happen if there are some hold in `registeredId2ClassInfo`? For example:
> 
> ```java
> // registeredClassIdCounter = 99
> fury.register(class1); // registeredClassIdCounter = 100 now
> fury.register(class2, 101); // registeredClassIdCounter = 101 now
> fury.register(class3); // throw exception since 101 has been used
> ```

I think for `org.apache.fury.resolver.ClassResolver#register(java.lang.Class<?>)` , it should always succeed. The implemention shoukd handle class id conflict inside

Reply:
> Yes, registered class id should be exposed to users, so they can know which ids they could use to avoid the conflict

We can abstract a separate class for ClassID allocation, provide an interface for users to obtain ClassID, and then reference this class in `ClassResolver` (Currently, the ClassResolver class is responsible for a lot of content, and this can also reduce the burden on the `ClassResolver` class).

In this way, we have one more class(For example named `ClassIdAllocator`) responsible for maintaining ClassID allocation, which can be used internally (`ClassResolver`) or externally (`Fury#allocateClassId`).

Reply:
> > What will happen if there are some hold in `registeredId2ClassInfo`? For example:
> > ```java
> > // registeredClassIdCounter = 99
> > fury.register(class1); // registeredClassIdCounter = 100 now
> > fury.register(class2, 101); // registeredClassIdCounter = 101 now
> > fury.register(class3); // throw exception since 101 has been used
> > ```
> 
> I think for `org.apache.fury.resolver.ClassResolver#register(java.lang.Class<?>)` , it should always succeed. The implemention shoukd handle class id conflict inside

Therefore, there is no need to deal with Classd conflicts in `org.apache.fury.resolver.ClassResolver#register(java.lang.Class<?>)`, because there will be no conflicts in ClassId allocated using `ClassIdAllocator`

Reply:
If you think there is nothing wrong with abstracting a `ClassIdAllocator`, and there is nothing unusual about the above suggestions, I will code and implement a `ClassIdAllocator` in the next few days. @chaokunyang

Reply:
Could you elabrate more the `ClassIdAllocator`? 
I'm not sure how it handle such an situation: there are three libraries and one client of three libraries. Every libraries has some classes to register, but they can't assume the register order since they don't know which will be registered first, so they use specified class id for such registration. If the client has two versions, the first version use library1 and library2, and the second version use library1 and library3. Then the registered ids in library2 should be left here, those ids shouldn't be used by library3. Will the `ClassIdAllocator` handle such cases?

Reply:
I agree with you, the current `ClassResolver` is a little complex, it would be great if we can refactor some code out.

Reply:
Can we create a new issue to discuss `ClassIdAllocator`? I think the content related to `ClassIdAllocator` should not be related to the current PR.

Maybe we can merge this PR first.

Reply:
Yes, could you open an issue for `ClassIdAllocator`? For this PR, maybe we still need to ensure `Fury.register` doesn't raise exception for id conflict

Reply:
`registeredClassIdCounter` maybe can be renamed to `classIdGenerator` for better intuition

Reply:
I have created an issue and expressed my thoughts on the question you raised above. https://github.com/apache/incubator-fury/issues/1412